File: src/frontend/src/api/memory.ts
export interface ShortTermMemoryResponse {
  thread_id: string;
  recent_messages: Array<{ role: string; content: string }>;
  conversation_summary: string | null;
  open_work_items: Record<string, unknown>[];
  pending_proposals: Record<string, unknown>[];
  task_dag: Record<string, unknown>[];
  task_batches: string[][];
  artifact_summaries: Record<string, unknown>[];
  action_results: Record<string, unknown>[];
}

export interface MemoryRecordResponse {
  id: string;
  namespace: string;
  memory_key: string;
  memory_type: string;
  content: Record<string, unknown>;
  confidence: number;
  status: string;
  source_thread_id: string | null;
  source_message_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface MarkdownExportResponse {
  files: string[];
}

// 阶段 9 的 memory 客户端只访问后端事实来源。前端不自行决定长期记忆
// 是否生效，避免把临时聊天指令写成长期偏好。
export async function fetchShortTermMemory(threadId: string): Promise<ShortTermMemoryResponse> {
  const response = await fetch(`/api/memory/threads/${encodeURIComponent(threadId)}/short-term`);

  if (!response.ok) {
    throw new Error(`Fetch short-term memory failed with ${response.status}`);
  }

  return response.json() as Promise<ShortTermMemoryResponse>;
}

export async function createMemoryCandidate(
  threadId: string,
  message: string,
): Promise<MemoryRecordResponse | null> {
  const response = await fetch("/api/memory/candidates", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ thread_id: threadId, message }),
  });

  if (!response.ok) {
    throw new Error(`Create memory candidate failed with ${response.status}`);
  }

  return response.json() as Promise<MemoryRecordResponse | null>;
}

export async function exportMarkdownBundle(): Promise<MarkdownExportResponse> {
  const response = await fetch("/api/memory/exports/markdown", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ include_audit: true }),
  });

  if (!response.ok) {
    throw new Error(`Export markdown failed with ${response.status}`);
  }

  return response.json() as Promise<MarkdownExportResponse>;
}

File: src/backend/app/schemas/memory.py
from typing import Any

from pydantic import BaseModel, Field


class RecentMessage(BaseModel):
    """短期记忆中的最近消息。"""

    role: str
    content: str


class ShortTermMemoryResponse(BaseModel):
    """短期记忆响应。

    该结构面向 Prompt 组装和前端调试，避免把完整 checkpoint 原样注入模型。
    """

    thread_id: str
    recent_messages: list[RecentMessage]
    conversation_summary: str | None = None
    open_work_items: list[dict[str, Any]]
    pending_proposals: list[dict[str, Any]]
    task_dag: list[dict[str, Any]]
    task_batches: list[list[str]]
    artifact_summaries: list[dict[str, Any]]
    action_results: list[dict[str, Any]]


class MemoryCandidateCreateRequest(BaseModel):
    """创建长期记忆候选的请求。"""

    thread_id: str | None = None
    message: str = Field(min_length=1)
    namespace: str = "preferences"


class MemoryRecordResponse(BaseModel):
    """长期记忆记录响应。"""

    id: str
    namespace: str
    memory_key: str
    memory_type: str
    content: dict[str, Any]
    confidence: float
    status: str
    source_thread_id: str | None = None
    source_message_id: str | None = None
    created_at: str
    updated_at: str


class ContactNoteResponse(BaseModel):
    """联系人备注召回响应。"""

    contact_email: str
    notes: list[MemoryRecordResponse]


class MarkdownExportRequest(BaseModel):
    """Markdown 导出请求。"""

    include_audit: bool = True


class MarkdownExportResponse(BaseModel):
    """Markdown 导出响应。"""

    files: list[str]
